staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -166,6 +166,14 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("listKind"), names.ListKind, "kind and listKind may not be the same"))
 	}
 
+	// kind cannot be a top-level type since they are shared across all groupversions
+	topLevelTypes := []string{"ListOptions", "ExportOptions", "GetOptions", "DeleteOptions", "Status", "APIVersions", "APIGroupList", "APIGroup", "APIResourceList", "WatchEvent"}
```

Comment:
a few questions:
* what keeps this from drifting?
* if we add a new meta type, does that mean we invalidate existing CRD objects?
* should this validation be done during creation, not update (to allow invalid named CRDs to be deleted)?

Reply:
> if we add a new meta type, does that mean we invalidate existing CRD objects?

We have to stop using these types in every api group, but start using meta.k8s.io/v1. Compare https://github.com/kubernetes/kubernetes/blob/b5b62c68318be79a665257c260ea9f9bbb6d6318/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go#L43.

Reply:
Can you check against unversioned types in a scheme and the usual meta types added via metav1.AddToScheme?

Reply:
> Can you check against unversioned types in a scheme and the usual meta types added via metav1.AddToScheme?

Done. It checks all the types in the scheme with group = "" and version = v1, since that's the group version for `UnversionedTypes` and for the meta types added via `metav1.AddToGroupVersion`:

https://github.com/kubernetes/kubernetes/blob/049b76201a76b8157679904d8d183546bc6a8a5a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L76-L77

Reply:
> what keeps this from drifting?

It uses the types in the scheme now.

> if we add a new meta type, does that mean we invalidate existing CRD objects?

If it's also added to the scheme, yes...

> should this validation be done during creation, not update (to allow invalid named CRDs to be deleted)?

@liggitt Do you mean this to allow already invalid CRDs to exist?